platforms/core-configuration/model-core/src/main/java/org/gradle/api/internal/provider/EvaluationContext.java

```diff
@@ -0,0 +1,269 @@
+     * @param evaluation the evaluation
+     * @param <R> the type of the result
+     * @param <E> (optional) exception type being thrown by the evaluation
+     * @return the result of the evaluation
+     * @throws E exception from the {@code evaluation} is propagated
+     */
+    public <R, E extends Exception> R tryEvaluate(ProviderInternal<?> provider, R fallbackValue, ScopedEvaluation<? extends R, E> evaluation) throws E {
+        if (getContext().isInScope(provider)) {
+            return fallbackValue;
+        }
+        // It is possible that the downstream chain itself forms a cycle.
+        // However, it should be its responsibility to be defined in terms of safe evaluation rather than us intercepting the failure here.
+        return evaluate(provider, evaluation);
+    }
+
+    /**
+     * Runs the {@code evaluation} in a nested evaluation context. A nested context allows to re-enter evaluation of the providers that are being evaluated in the enclosed context.
+     * <p>
+     * Use sparingly. In most cases, it is better to rework the call chain to avoid re-evaluating the provider.
+     *
+     * @param evaluation the evaluation
+     * @param <R> the type of the result
+     * @param <E> (optional) exception type being thrown by the evaluation
+     * @return the result of the evaluation
+     * @throws E exception from the {@code evaluation} is propagated
+     */
+    public <R, E extends Exception> R evaluateNested(ScopedEvaluation<? extends R, E> evaluation) throws E {
+        try (ScopeContext ignored = nested()) {
+            return evaluation.evaluate();
+        }
+    }
+
+    private PerThreadContext getContext() {
+        return threadLocalContext.get();
+    }
+
+    private PerThreadContext setContext(PerThreadContext newContext) {
+        threadLocalContext.set(newContext);
+
+        return newContext;
+    }
+
+    private ScopeContext nested() {
+        return setContext(new PerThreadContext(getContext()));
+    }
+
+    private final class PerThreadContext implements ScopeContext {
+        private final Set<ProviderInternal<?>> providersInScope = new ReferenceOpenHashSet<>(EXPECTED_MAX_CONTEXT_SIZE);
+        private final List<ProviderInternal<?>> providersStack = new ReferenceArrayList<>(EXPECTED_MAX_CONTEXT_SIZE);
+        @Nullable
+        private final PerThreadContext parent;
+
+        public PerThreadContext(@Nullable PerThreadContext parent) {
+            this.parent = parent;
+        }
+
+        private void push(ProviderInternal<?> provider) {
+            if (providersInScope.add(provider)) {
+                providersStack.add(provider);
+            } else {
+                throw prepareException(provider);
+            }
+        }
+
+        private void pop() {
+            ProviderInternal<?> removed = providersStack.remove(providersStack.size() - 1);
+            providersInScope.remove(removed);
+        }
+
+        public PerThreadContext enter(ProviderInternal<?> owner) {
```

Comment:
Maybe `open` (to match `close()`), or make `AutoCloseable#close()` an on-liner that calls `leave()` or `exit()` (to match `enter()`).

💭